x/supplier/keeper/msg_server_submit_proof.go

```diff
@@ -4,55 +4,130 @@ import (
 	"context"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 
-	"github.com/pokt-network/poktroll/x/supplier/types"
+	suppliertypes "github.com/pokt-network/poktroll/x/supplier/types"
 )
 
-func (k msgServer) SubmitProof(goCtx context.Context, msg *types.MsgSubmitProof) (*types.MsgSubmitProofResponse, error) {
+func (k msgServer) SubmitProof(goCtx context.Context, msg *suppliertypes.MsgSubmitProof) (*suppliertypes.MsgSubmitProofResponse, error) {
 	// TODO_BLOCKER: Prevent Proof upserts after the tokenomics module has processes the respective session.
 	// TODO_BLOCKER: Validate the signature on the Proof message corresponds to the supplier before Upserting.
-
 	ctx := sdk.UnwrapSDKContext(goCtx)
+	logger := k.Logger(ctx).With("method", "SubmitProof")
+	logger.Debug("submitting proof")
+
+	/*
+		TODO_INCOMPLETE: Handling the message
+
+		## Actions (error if anything fails)
+		1. Retrieve a fully hydrated `session` from on-chain store using `msg` metadata
+		2. Retrieve a fully hydrated `claim` from on-chain store using `msg` metadata
+		3. Retrieve `relay.Req` and `relay.Res` from deserializing `proof.ClosestValueHash`
+
+		## Basic Validations (metadata only)
+		1. proof.sessionId == claim.sessionId
+		2. msg.supplier in session.suppliers
+		3. relay.Req.signer == session.appAddr
+		4. relay.Res.signer == msg.supplier
+
+		## Msg distribution validation (governance based params)
+		1. Validate Proof submission is not too early; governance-based param + pseudo-random variation
+		2. Validate Proof submission is not too late; governance-based param + pseudo-random variation
+
+		## Relay Mining validation
+		1. verify(proof.path) is the expected path; pseudo-random variation using on-chain data
+		2. verify(proof.ValueHash, expectedDiffictulty); governance based
+		3. verify(claim.Root, proof.ClosestProof); verify the closest proof is correct
+	*/
 
 	if err := msg.ValidateBasic(); err != nil {
-		return nil, err
+		return nil, status.Error(codes.InvalidArgument, err.Error())
```

Comment:
This seems like a pattern you have found, is it worth updating the codebase **post migration** to do this for all message handlers? Should we create an issue for this?

It looks like it stops us from using `require.ErrorIs` / `require.ErrorAs` so some tests will need to be updated too.

Reply:
The CLI should ALWAYS return a gRPC status error (i.e. `status.Error(...)`), in addition to consistency, this is an important aspect of separating client-facing errors from internal errors. Details of internal errors should only appear in logs that the operator sees and NEVER be sent to the client to avoid leaking potentially sensitive information.

This pattern fell out of the refactoring that was necessary in #266/#250. The tests in those branches take this into consideration (and will be re-applied in subsequent work).

IIRC the reason I pushed usage of `status.Error` all the way down into the keeper was because there was a lot of redundancy appearing in the CLI code involving wrapping keeper errors in the same status code error. Given that the only consumer of the keepers should be the CLI, or one another, it didn't seem inappropriate to make that simplification. I don't have a strong opinion about it though.